### ISSUE TITLE
twisted.web.proxy incorrectly mangles URIs

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3013]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3013|
|Type|defect|
|Created|2008-01-25 22:17:13Z|
|Branch|https://github.com/twisted/twisted/tree/proxy-urlquote-3013|

`ReverseProxyResource.render` parses the request URI it receives.  It then passes on the path and the query portions to the target of the proxying.

It does not pass on the fragment.

It unquotes the URI and does not re-quote it when sending it on.

<details><summary>Searchable metadata</summary>

```
trac-id__3013 3013
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__branches_proxy_urlquote_3013 branches/proxy-urlquote-3013
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1201299433000000 1201299433000000
changetime__1201793275000000 1201793275000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Huh, actually, I have no idea what the correct behavior WRT the fragment is.  The quoting issue is real and serious, though.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [22401]) Branching to 'proxy-urlquote-3013'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="danderson's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|danderson commented|
|-|-|

The changes on the branch look good.

As for the fragment question, I don't think that a normal HTTP client would ever send a fragment to a server, since it's just a local referencing system, none of the server's business. At least Firefox and wget filter out the fragment part of the URL before sending the request to the server. Telnetting to an Apache server and GETing an URL with a fragment returns the same resource as the URL without fragment, but the fragment is made available in logs, and to scripts (eg. php).

Given all that, it seems that the issue is moot for most applications, since they don't actually send a fragment. But, to make the proxy as transparent as posible, I think the fragment should be passed on. Let the destination server decide what to do with it.

**Comment by twisted-trac** [comment]
|<img alt="danderson's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|danderson set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun removed owner|
|-|-|

Thanks for the review.

Passing on the fragment makes sense to me, but it's going to be pretty difficult to implement.

`ReverseProxyResource` learns about the URI through the `getChild` API.  The segment passed to `getChild` is already decoded.  This means there's no way to differentiate between the `#` which delimits the fragment and a `#` which was really a `%23` in the request.  So, while ideally I'd change this code to pass on fragments, I'm not actually excited about changing the whole proxying implementation in order to support this non-standard feature.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to @exarkun|
|-|-|

Why did you mangle the quote import as urlquote?

Otherwise, looks good.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

```
15:06 < exarkun> therve: "quote" seems too vague to me
15:07 < therve> exarkun: alright
15:07 < therve> exarkun: you can merge then
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set status to `closed`|
|-|-|

(In [22431]) Merge proxy-urlquote-3013

Author: exarkun
Reviewer: danderson, therve
Fixes [#3013](https://github.com/twisted/twisted/issues/3013)

Change twisted.web.proxy.ReverseProxyResource so that it correctly encodes
the request URI it sends on to the server for which it is a proxy.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
